proposals/p3980.md

```diff
@@ -0,0 +1,398 @@
+-   [Open discussion 2024-05-09: Number of allowed redeclarations](https://docs.google.com/document/d/1s3mMCupmuSpWOFJGnvjoElcBIe2aoaysTIdyczvKX84/edit?resourcekey=0-G095Wc3sR6pW1hLJbGgE0g&tab=t.0#heading=h.bu7djkos4xo)
+-   [Issue #3986: Alternative naming for `has_extern` keyword](https://github.com/carbon-language/carbon-lang/issues/3986)
+
+## Proposal
+
+A given entity may have up to three declarations:
+
+-   An optional `extern` declaration
+    -   It must be in a separate library from the definition.
+    -   The owning library's API file must import the `extern` declaration, and
+        must also contain a declaration.
+-   An optional forward declaration
+    -   This must come before the definition. The API file is considered to be
+        before the implementation file.
+-   A required definition
+
+The first owning declaration must have the `has_extern` modifier.
+
+The consequential changes to the [problem example](#problem) are then:
+
+```
+library "a";
+
+// This proposal makes the import required.
+import library "b";
+
+// This proposal adds `has_extern`.
+has_extern class C {}
+```
+
+```
+library "b";
+extern class C;
+extern fn F() -> C*;
+```
+
+```
+library "c";
+import library "a";
+extern fn F() -> C*;
+```
+
+## Details
+
+### Type coherency
+
+In the context of the example that is the [problem](#problem), `C` will produce
+the same type regardless of whether `C` is the `extern` or non-`extern`
+declaration. This means that both function signatures have identical types.
+
+We do this by having the non-`extern` declaration import the `extern`
+declaration. Because one can see the other, the compiler can easily use the same
+type for both declarations. This makes it easy for compilation of other
+libraries, which may be importing either or both declarations, to easily
+determine that types are equivalent.
+
+### Using imported declarations
+
+Since `extern class C;` must be imported by the owning library, we now allow
+uses of the imported name prior to its declaration within the same file. This is
+a divergence from
+[#3762](https://github.com/carbon-language/carbon-lang/pull/3762). It means the
+following now works:
+
+```
+library "extern";
+
+extern class MyType;
+```
+
+```
+library "use_extern";
+import library "extern"
+
+// Uses the `extern` declaration.
+fn Foo(val: MyType*);
+
+has_extern class MyType {
+  fn Bar[addr self: Self*]() { Foo(self); }
+}
+```
+
+### `has_extern` modifier
+
+The `has_extern` modifier must be present on the first owning declaration, when
+there is an `extern` declaration. Because the first owning declaration must be
+in an API file, this will only be present in API files.
+
+The modifier allows libraries to restrict whether they have an `extern`
+declaration. It will also be used to support tool-based validation that the
+`extern` declaration is imported as required.
+
+### Versus current state
```

Comment:
One thing I'm unclear on here, mostly because I'm not entirely sure what the current state is:

Given
```
library "extern";
extern class C;
fn F() -> C;
```
```
library "owning";
import "extern";
has_extern class C {}
```
```
library "distant";
import "owning";
```
```
library "client";
import "extern";
import "distant";
fn CallF() {
  F();
}
```
... is the call to `F()` in `CallF()` valid? That is: is `C` a complete type in that compilation?

If so, then removing the `import "owning";` from library `"distant"` breaks the build of the `"client"` library. That seems problematic.

If not, then presumably that only applies for `has_extern` types. That is, I think we want this to work:
```
library "owning";
class C {}
```
```
library "distant";
import "owning";
fn F() -> C;
```
```
library "client";
import "distant";
fn CallF() {
  F();
}
```
... and I think the relevant difference here is whether `C` is known to have `extern` declarations.

A rule we've previously discussed, but I don't remember whether we've agreed to or not, is:

- Definitions of types are transitively visible (even if not explicitly re-exported) if they are not `has_extern`
- Definitions of `has_extern` types are not transitively visible, and those types are instead incomplete if their owning library (or a library that `export name;`s the type, or maybe `export name;`s something involving the type?) is not directly imported.

I'm fine saying this is off-topic for this PR, but if so, it'd be useful to at least mention what the current state is -- if we don't have such a rule, then capturing that there is a known issue with removing distant imports in this PR would be useful.

Reply:
> ... is the call to `F()` in `CallF()` valid? That is: is `C` a complete type in that compilation?

`C` is never imported by `client` so no. There is no reason its type would be complete.

It might be different if something pulled the complete definition in for `client`.

> If so, then removing the `import "owning";` from library `"distant"` breaks the build of the `"client"` library. That seems problematic.

N/A, although you might get interactions such as this, just not from the example. e.g., if you also had something like your later example used in combination, which triggered the definition of `C` to be pulled in.

> If not, then presumably that only applies for `has_extern` types. That is, I think we want this to work:
> 
[elided]
> 
> ... and I think the relevant difference here is whether `C` is known to have `extern` declarations.

Yes, because this dealing with a complete type, and we have only one type.

> A rule we've previously discussed, but I don't remember whether we've agreed to or not, is:
> 
> * Definitions of types are transitively visible (even if not explicitly re-exported) if they are not `has_extern`
> * Definitions of `has_extern` types are not transitively visible, and those types are instead incomplete if their owning library (or a library that `export name;`s the type, or maybe `export name;`s something involving the type?) is not directly imported.
> 
> I'm fine saying this is off-topic for this PR, but if so, it'd be useful to at least mention what the current state is -- if we don't have such a rule, then capturing that there is a known issue with removing distant imports in this PR would be useful.

I think declarations are transitively visible regardless of `has_extern`. The only difference with `has_extern` is whether declarations being transitively visible might have significant effect.

Note, I think this kind of thing is an unavoidable consequence of type consolidation: we can't expect `extern` to be using the same type as `has_extern`, but have the same type behave differently in use.

Reply:
Discussing the awkward consequences of `has_extern` in the combination case, in particular if you have code that imports functions which return `has_extern` and `extern` versions of the type (indirectly imported, and with extra consideration cross-package where names are lazy loaded), I think there are three main options:

1. The order of calling functions matters, under the information accumulation principle. If you call the `has_extern` version first, the type is complete when the `extern` version is called. In the other order, it's still incomplete.

2. Whenever a library has an exported entity which uses a `has_extern` type, it adds that `has_extern` type to a list of types which _must_ be evaluated on import. This is not transitive unless re-used. Thus call order doesn't matter, and a complete type is available.

    - There's a simpler option here where all imported `has_extern` can be treated this way, but the restriction to exported entities may simplify matters. Although in turn, it may become really complex for class members.

3. If a `has_extern` type is indirectly returned (and not directly imported) the type is always incomplete.

    - I think this is the rule you're describing.

Reply:
Filed #4025